### PR TITLE
Add checkpoints

### PR DESCRIPTION
**Pull request by SoluMilken**
In this PR, `Pipe` has extra ability to collect intermediate result during transformation by using `add_checkpoint` and  `Intermediate`.
1. `add_checkpoint`:
This is an extra method of `Pipe` created in this PR. It appends the index of step into `self._checkpoints` to memorize which outcome of step is required to be stored.
For example,  if you create a Pipe as follow, you means to keep the outcome of operator2 and operator3 during transformation.
```python
p = Pipe()
p.add('op1_name', {})
p.add('op2_name', {}, checkpoint='0')
p.add('op3_name', {}, checkpoint='1')
p.add('op4_name', {})
```

2. `Intermediate`:
`Intermediate` is a class for collecting all outcomes of `Step`(or `Operator`) during transformation of `Pipe`. We involve `Intermediate` in the transformation process of `Pipe`. So,  `Pipe` will return an instance of `Intermediate` when `transform` is called. 
Take the above pipe as an example:
```python
>>> indices, intent_labels, entity_labels, realigner, intermediate = p.transform(datum)
>>> intermediate.get_from_checkpoint('0')
# return the result of Op2
>>> intermediate.get_from_checkpoint('1')
# return the result of Op3
>>> intermediate[0]
# return raw input
>>> intermediate[1]
# return the result of Op1  
```

### CODE REVIEW

File: uttut/pipeline/intermediate.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    def add(self, intermediate):
+        """Append intermediate into self.collection
+
+        Arg:
+            intermediate: object to be stored in self.collection
+
+        """
+        self.collection.append(intermediate)
+
+    def get(self, index: int = 0):
```
Comment (stegben) [inline uttut/pipeline/intermediate.py]: 私心還是覺得要提示完整一點，比如說 `get_from_checkpoint` 之類的，但這樣又有點冗。所以就看其他人有什麼想法

Comment (GBLin5566) [reply uttut/pipeline/intermediate.py]: I prefer ```get_from_checkpoint``` too. It's confusing that ``` get ``` & ``` __getitem__ ``` have different behaviors.

Comment (SoluMilken) [reply uttut/pipeline/intermediate.py]: @absolutelyNoWarranty

Comment (absolutelyNoWarranty) [reply uttut/pipeline/intermediate.py]: +1 `get_from_checkpoint`

File: uttut/pipeline/intermediate.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        """Query intermediate in collection according to input index
+
+        Arg:
+            index (int)
```
Comment (absolutelyNoWarranty) [inline uttut/pipeline/intermediate.py]:
```suggestion
            index (int): index of checkpoint
```

**Comment by SoluMilken** [comment]
修改過得README
```python
>>> from uttut.pipeline.pipe import Pipe

>>> p = Pipe()
>>> p.add('IntTokenWithSpace')
>>> p.add('FloatTokenWithSpace')
>>> p.add('MergeWhiteSpaceCharacters')
>>> p.add('StripWhiteSpaceCharacters')
>>> p.add('EngTokenizer')  # word-level (ref: BERT)
>>> p.add('AddSosEos')
>>> p.add_checkpoint()
>>> p.add('Pad', {'maxlen': 5})
>>> p.add(
    'Token2Index',
    {
       'token2index': {
            '<sos>': 0, '<eos>': 1,  # for  AddSosEos
            '<unk>': 2, '<pad>': 3,  # for Pad
            '_int_': 4,  # for IntTokenWithSpace
            '_float_': 5,  # for FloatTokenWithSpace
            'I': 6,
            'apples': 7,
        },
    },
)

>>> from uttut.elements import Datum, Entity, Intent
>>> datum = Datum(
    utterance='I like apples.',
    intents=[Intent(label=1), Intent(label=2)],
    entities=[Entity(start=7, end=13, value='apples', label=7)],
)
```

**Comment by SoluMilken** [comment]
@absolutelyNoWarranty 還要改一些東西 請大家不要 merge 這個 PR

File: uttut/pipeline/intermediate.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    attributes:
+        collection (list): store input intermediates
+        checkpoints (ints): indices of intermediates for query
+
+    """
+
+    def __init__(self, checkpoints: List[int]):
+        self.collection: list
+        self.collection = []
```
[the remark this answers is not in the document]
Comment (SoluMilken) [reply uttut/pipeline/intermediate.py]: 7282023

File: uttut/pipeline/intermediate.py (hunk 1 above)
Comment (SoluMilken) [reply uttut/pipeline/intermediate.py]: 7282023

**Pull request merged**